File: android/app/src/main/java/com/privycs/vpn/license/License.kt
package com.privycs.vpn.license

import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import org.json.JSONObject
import java.security.MessageDigest

/**
 * Privycs cross-platform Pro license-key verifier.
 *
 * Byte-identical wire format to the Desktop verifier
 * (desktop/internal/license/license.go) and the gateway-side LemonSqueezy
 * webhook signer (when wired). The same `.privycs-license` file activates
 * on Android, Desktop and (later) iOS — the bundle SKU's promise.
 *
 * Format:
 *
 *     PRVC-{base32(canonicalJSON(payload))}-{base32(ed25519-sig)}
 *
 * base32 = RFC 4648 standard alphabet, no padding, uppercase.
 *
 * Verification is OFFLINE — the embedded public key (BuildConfig
 * .LICENSE_PUBLIC_KEY_HEX) is the trust anchor. No phone-home, no online
 * validation; an offline-only customer can activate.
 *
 * Stdlib + BouncyCastle only. No coroutines, no Android dependencies in
 * this file — pure Kotlin/JVM so the verifier is unit-testable on the
 * JVM without an emulator.
 */
object License {
    const val PREFIX = "PRVC"
    private const val SEP = "-"

    const val SKU_DESKTOP = "privycs_pro_desktop"
    const val SKU_BUNDLE = "privycs_pro_bundle_all"

    const val PLATFORM_ANDROID = "android"
    const val PLATFORM_DESKTOP = "desktop"
    const val PLATFORM_IOS = "ios"

    /** Schema version this app accepts. Bumping requires a coordinated
     *  signer + all-platform verifier upgrade. */
    const val CURRENT_SCHEMA_VERSION = 1

    /**
     * Payload structure inside the canonical-JSON segment. Field names
     * (JSON keys) MUST match the signer exactly — the verifier reads via
     * org.json with explicit key names below.
     */
    data class Payload(
        val v: Int,
        val tier: String,
        val sku: String,
        val platforms: List<String>,
        val issued: String,
        val buyerEmailHash: String?,
    )

    /** Distinct error kinds so the UI can render a tailored message. */
    enum class ErrorKind {
        MALFORMED,
        BAD_SIGNATURE,
        UNSUPPORTED_VERSION,
        WRONG_PLATFORM,
        UNKNOWN_TIER,
        NO_PUBLIC_KEY,
    }

    sealed class Result {
        data class Ok(val payload: Payload) : Result()
        data class Err(val kind: ErrorKind, val detail: String) : Result()
    }

    /**
     * Verify a raw PRVC-...-... key against [pubKeyHex] (a 64-char hex
     * string of the 32-byte ed25519 public key). Returns Ok with the
     * parsed payload on success.
     *
     * The caller passes [expectedPlatform] (typically [PLATFORM_ANDROID])
     * — keys whose payload.platforms array doesn't include it are
     * rejected with WRONG_PLATFORM. Cross-platform bundle keys list all
     * three platforms; a single-Desktop key would fail here as expected.
     */
    fun verify(
        rawKey: String,
        pubKeyHex: String,
        expectedPlatform: String = PLATFORM_ANDROID,
    ): Result {
        if (pubKeyHex.isEmpty()) {
            return Result.Err(ErrorKind.NO_PUBLIC_KEY, "no pubkey configured")
        }
        val pubKey = hexToBytes(pubKeyHex)
            ?: return Result.Err(ErrorKind.NO_PUBLIC_KEY, "invalid hex")
        if (pubKey.size != 32) {
            return Result.Err(ErrorKind.NO_PUBLIC_KEY, "wrong length ${pubKey.size}")
        }

        val parts = rawKey.trim().split(SEP)
        if (parts.size != 3 || parts[0] != PREFIX) {
            return Result.Err(ErrorKind.MALFORMED, "expected PRVC-<payload>-<sig>")
        }

        val payloadBytes = decodeBase32(parts[1])
            ?: return Result.Err(ErrorKind.MALFORMED, "payload not base32")
        val sigBytes = decodeBase32(parts[2])
            ?: return Result.Err(ErrorKind.MALFORMED, "sig not base32")
        if (sigBytes.size != 64) {
            return Result.Err(ErrorKind.MALFORMED, "sig length ${sigBytes.size}, want 64")
        }

        if (!ed25519Verify(pubKey, payloadBytes, sigBytes)) {
            return Result.Err(ErrorKind.BAD_SIGNATURE, "signature does not match")
        }

        val payload = parsePayload(payloadBytes)
            ?: return Result.Err(ErrorKind.MALFORMED, "payload not JSON")
        if (payload.v != CURRENT_SCHEMA_VERSION) {
            return Result.Err(ErrorKind.UNSUPPORTED_VERSION, "v=${payload.v}")
        }
        if (payload.tier != "pro") {
            return Result.Err(ErrorKind.UNKNOWN_TIER, payload.tier)
        }
        if (expectedPlatform.isNotEmpty() && expectedPlatform !in payload.platforms) {
            return Result.Err(
                ErrorKind.WRONG_PLATFORM,
                "platforms=${payload.platforms} does not include $expectedPlatform",
            )
        }
        return Result.Ok(payload)
    }

    /** SHA-256 hex of lowercased trimmed email — matches the Go signer. */
    fun hashBuyerEmail(email: String): String {
        val md = MessageDigest.getInstance("SHA-256")
        val digest = md.digest(email.trim().lowercase().toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    // ---- private helpers ----

    private fun ed25519Verify(pubKey: ByteArray, message: ByteArray, signature: ByteArray): Boolean {
        return try {
            val signer = Ed25519Signer()
            signer.init(false, Ed25519PublicKeyParameters(pubKey, 0))
            signer.update(message, 0, message.size)
            signer.verifySignature(signature)
        } catch (_: Throwable) {
            false
        }
    }

    /**
     * Parse the canonical-JSON payload bytes into a [Payload]. Uses
     * org.json (built into Android) — we don't need the canonicalisation
     * here because the signer already produced canonical bytes and the
     * verifier matches the SIGNATURE against those exact bytes. JSON
     * decoding is for the field-level access only.
     */
    private fun parsePayload(bytes: ByteArray): Payload? {
        return try {
            val obj = JSONObject(String(bytes, Charsets.UTF_8))
            val platforms = mutableListOf<String>()
            val arr = obj.optJSONArray("platforms")
            if (arr != null) {
                for (i in 0 until arr.length()) {
                    platforms.add(arr.getString(i))
                }
            }
            Payload(
                v = obj.optInt("v", 0),
                tier = obj.optString("tier", ""),
                sku = obj.optString("sku", ""),
                platforms = platforms.toList(),
                issued = obj.optString("issued", ""),
                buyerEmailHash = obj.optString("buyer_email_hash").takeIf { it.isNotEmpty() },
            )
        } catch (_: Throwable) {
            null
        }
    }

    /**
     * RFC 4648 base32 decoder — standard alphabet, no padding, uppercase.
     * Matches Go's `base32.StdEncoding.WithPadding(base32.NoPadding)` so
     * round-trip with the signer is byte-identical. Returns null on
     * invalid character.
     */
    private fun decodeBase32(s: String): ByteArray? {
        val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        val src = s.uppercase()
        if (src.isEmpty()) return ByteArray(0)
        val out = ByteArray(src.length * 5 / 8)
        var buffer = 0L
        var bitsLeft = 0
        var outIdx = 0
        for (c in src) {
            val v = alphabet.indexOf(c)
            if (v < 0) return null
            buffer = (buffer shl 5) or v.toLong()
            bitsLeft += 5
            if (bitsLeft >= 8) {
                bitsLeft -= 8
                if (outIdx >= out.size) return null
                out[outIdx++] = ((buffer shr bitsLeft) and 0xFF).toByte()
            }
        }
        return out.copyOf(outIdx)
    }

    private fun hexToBytes(s: String): ByteArray? {
        if (s.length % 2 != 0) return null
        return try {
            ByteArray(s.length / 2) { i ->
                Integer.parseInt(s.substring(i * 2, i * 2 + 2), 16).toByte()
            }
        } catch (_: NumberFormatException) {
            null
        }
    }
}

File: desktop/internal/license/license.go
// Package license implements the Privycs ed25519-signed offline
// license-key format used by all 3 client platforms (Desktop / Android
// / iOS). It is intentionally Go-stdlib only — both the desktop client
// (verify) and the cmd/license-gen dev tool (sign) consume identical
// canonical-JSON bytes, so any subtle drift between signer and
// verifier would break activation.
//
// On-wire format:
//
//	PRVC-{base32(canonicalJSON(payload))}-{base32(ed25519-sig)}
//
// See ../../entitlement_crypto.go for the desktop-side wrapper that
// pulls in the embedded public key, and ../../../cmd/license-gen for
// the dev signing tool.
package license

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base32"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"slices"
	"sort"
	"strings"
	"time"
)

const (
	// Prefix is the literal first segment all valid keys share.
	Prefix = "PRVC"
	// Sep is the segment separator inside a key.
	Sep = "-"

	// SKUDesktop is the single-Desktop product code on LemonSqueezy.
	SKUDesktop = "privycs_pro_desktop"
	// SKUBundle is the cross-platform bundle product code (Android +
	// Desktop + iOS in one key).
	SKUBundle = "privycs_pro_bundle_all"

	// PlatformDesktop / PlatformAndroid / PlatformIOS are the values
	// the Platforms array in a payload uses. A client only accepts a
	// key whose Platforms array contains its own platform.
	PlatformDesktop = "desktop"
	PlatformAndroid = "android"
	PlatformIOS     = "ios"

	// CurrentSchemaVersion is what we set on Issue and what Verify
	// rejects as "unsupported" if not matched. Bumping requires a
	// migration path because old apps cannot verify keys with newer
	// schemas.
	CurrentSchemaVersion = 1
)

// Payload is the signed portion of a license key. Field tags are
// alphabetic and match the order used by canonicalJSON's lexicographic
// re-encoding — signer and verifier produce identical bytes regardless
// of struct field order or marshaller quirks.
type Payload struct {
	BuyerEmailHash string   `json:"buyer_email_hash,omitempty"`
	Issued         string   `json:"issued"`
	Platforms      []string `json:"platforms"`
	SKU            string   `json:"sku"`
	Tier           string   `json:"tier"`
	V              int      `json:"v"`
}

// Public verify-side errors. Wrappable with errors.Is so the UI can
// render a specific message ("Wrong platform" vs "Bad signature").
var (
	ErrMalformed     = errors.New("license key malformed")
	ErrBadSignature  = errors.New("license signature invalid")
	ErrUnsupportedV  = errors.New("license schema version not supported by this app")
	ErrWrongPlatform = errors.New("license is not valid for this platform")
	ErrUnknownTier   = errors.New("license tier not recognised")
	ErrNoPublicKey   = errors.New("no public key configured")
)

// Verify parses a raw "PRVC-..." key and checks the ed25519 signature
// against pub. Returns the parsed payload on success. The caller is
// responsible for additionally checking that the payload's Platforms
// includes the caller's own platform (e.g. PlatformDesktop).
//
// Pure / re-entrant — no I/O, no globals.
func Verify(rawKey string, pub ed25519.PublicKey, expectedPlatform string) (*Payload, error) {
	rawKey = strings.TrimSpace(rawKey)
	parts := strings.Split(rawKey, Sep)
	if len(parts) != 3 || parts[0] != Prefix {
		return nil, fmt.Errorf("%w: expected %s-<payload>-<sig>", ErrMalformed, Prefix)
	}
	payloadBytes, err := DecodeBase32(parts[1])
	if err != nil {
		return nil, fmt.Errorf("%w: payload not base32: %v", ErrMalformed, err)
	}
	sigBytes, err := DecodeBase32(parts[2])
	if err != nil {
		return nil, fmt.Errorf("%w: sig not base32: %v", ErrMalformed, err)
	}
	if len(sigBytes) != ed25519.SignatureSize {
		return nil, fmt.Errorf("%w: sig length %d, want %d", ErrMalformed, len(sigBytes), ed25519.SignatureSize)
	}
	if len(pub) != ed25519.PublicKeySize {
		return nil, ErrNoPublicKey
	}
	if !ed25519.Verify(pub, payloadBytes, sigBytes) {
		return nil, ErrBadSignature
	}

	var pl Payload
	if err := json.Unmarshal(payloadBytes, &pl); err != nil {
		return nil, fmt.Errorf("%w: payload not JSON: %v", ErrMalformed, err)
	}
	if pl.V != CurrentSchemaVersion {
		return nil, fmt.Errorf("%w: v=%d", ErrUnsupportedV, pl.V)
	}
	if pl.Tier != "pro" {
		return nil, fmt.Errorf("%w: %q", ErrUnknownTier, pl.Tier)
	}
	if expectedPlatform != "" && !slices.Contains(pl.Platforms, expectedPlatform) {
		return nil, fmt.Errorf("%w: payload.platforms=%v does not include %q", ErrWrongPlatform, pl.Platforms, expectedPlatform)
	}
	return &pl, nil
}

// NOTE: Issue() — the ed25519 PRIVATE-key signer — has been moved out
// of the default build. It now lives in signer.go behind the
// `licensegen` build tag so the shipped client binary contains ONLY
// the verify side. The dev signing tool / webhook builds with
// `-tags licensegen` to pull it back in. See signer.go.

// CanonicalJSON serialises a Payload with keys in lexicographic order
// and no whitespace — the exact byte sequence that ed25519 signs and
// verifies against. Signer + verifier MUST use this same function;
// any drift (e.g. different marshaller) silently breaks activation.
func CanonicalJSON(pl *Payload) ([]byte, error) {
	b, err := json.Marshal(pl)
	if err != nil {
		return nil, err
	}
	var m map[string]any
	if err := json.Unmarshal(b, &m); err != nil {
		return nil, err
	}
	return marshalSorted(m), nil
}

// marshalSorted emits a JSON object with keys in lexicographic order.
// Recurses into nested objects; preserves array order (array order is
// part of the signed message — flipping ["android","desktop"] to
// ["desktop","android"] changes the signature).
func marshalSorted(m map[string]any) []byte {
	var b strings.Builder
	b.WriteByte('{')
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for i, k := range keys {
		if i > 0 {
			b.WriteByte(',')
		}
		writeJSONString(&b, k)
		b.WriteByte(':')
		writeJSONValue(&b, m[k])
	}
	b.WriteByte('}')
	return []byte(b.String())
}

func writeJSONValue(b *strings.Builder, v any) {
	switch x := v.(type) {
	case nil:
		b.WriteString("null")
	case bool:
		if x {
			b.WriteString("true")
		} else {
			b.WriteString("false")
		}
	case string:
		writeJSONString(b, x)
	case float64:
		// JSON numbers come back as float64 — re-encode integer
		// values without ".0" so they round-trip with the signer's
		// integer-typed payload (e.g. v=1, not v=1.0).
		if x == float64(int64(x)) {
			fmt.Fprintf(b, "%d", int64(x))
		} else {
			fmt.Fprintf(b, "%g", x)
		}
	case map[string]any:
		b.Write(marshalSorted(x))
	case []any:
		b.WriteByte('[')
		for i, item := range x {
			if i > 0 {
				b.WriteByte(',')
			}
			writeJSONValue(b, item)
		}
		b.WriteByte(']')
	default:
		raw, _ := json.Marshal(x)
		b.Write(raw)
	}
}

func writeJSONString(b *strings.Builder, s string) {
	raw, _ := json.Marshal(s)
	b.Write(raw)
}

// EncodeBase32 and DecodeBase32 use the RFC 4648 standard alphabet
// with NO padding so keys round-trip cleanly through email + UI fields
// without the visual clutter of trailing equal signs.
func EncodeBase32(b []byte) string {
	enc := base32.StdEncoding.WithPadding(base32.NoPadding)
	return enc.EncodeToString(b)
}

func DecodeBase32(s string) ([]byte, error) {
	enc := base32.StdEncoding.WithPadding(base32.NoPadding)
	return enc.DecodeString(strings.ToUpper(s))
}

// HashBuyerEmail returns the SHA-256 hex of the lowercased trimmed
// email. Stored as buyer_email_hash inside a payload so the LemonSqueezy
// support tooling can correlate a key to a purchase without leaking the
// email itself into the app or its UI.
func HashBuyerEmail(email string) string {
	h := sha256.Sum256([]byte(strings.ToLower(strings.TrimSpace(email))))
	return hex.EncodeToString(h[:])
}

// PayloadTimestamp returns the day-precision UTC string used for the
// Issued field. Centralised here so signer + UI + tests render the
// same format.
func PayloadTimestamp(t time.Time) string {
	return t.UTC().Format("2006-01-02")
}
